Improve 'no version property' feature

shestee opened the pull request:
1. When there is no version property configured and the project is being built not from tag, the plugin deducts version based on previous tag. This commit improves plugin's behaviour in such situation with increasing patch version number, taken from last tag, by 1 and adding `-SNAPSHOT` suffix. This solves #77

2. Additionally this commit fixes bug: when `version` property is not present and `tagPrefix` has no value set (`tagPrefix=`), current plugin version wrongly picks version based on tag but should pick `0.0.1`.  Also corrected unit test for behaviour when there is no config file and code is checked out on not valid tag.

shestee committed:
Improve 'no version property' feature

1. When there is no version property configured and the project is being built not
from tag, the plugin deducts version based on previous tag. This commit improves
plugin's behaviour in such situation with increasing patch version number,
taken from last tag, by 1 and adding '-SNAPSHOT' suffix.
2. When 'version' property is not present and 'tagPrefix' has no value set
('tagPrefix='), current plugin version wrongly picks version based on tag but
should pick '0.0.1' - this commit also fixes this bug.
Additionally corrected unit test for behaviour when no config file and not checked
out on valid tag.

mockitoguy commented on src/main/java/org/shipkit/auto/version/NextVersionPicker.java:
```diff
@@ -30,13 +34,18 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
+            if (isVersionTag(tag, config.getTagPrefix()) && isSupportedVersion(tag.substring(config.getTagPrefix().length()))) {
+                result = tag.substring(config.getTagPrefix().length());
+                explainVersion(log, result, "deducted version based on tag: '" + tag + "'");
+            } else if (isVersionTag(tag, config.getTagPrefix()) && isSnapshot(tag, config.getTagPrefix())){
```
Let's push the logic of `isVersionTag` to `isSnapshot` method. I think it will be cleaner. It's OK to do this in a separate PR. Bonus: we might do the same with `isSupportedVersion`.

mockitoguy commented on src/main/java/org/shipkit/auto/version/NextVersionPicker.java:
```diff
@@ -30,13 +34,18 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
+                Pattern pattern = Pattern.compile("\\d+\\.\\d+\\.\\d+");
+                Matcher matcher = pattern.matcher(tag);
+                matcher.find();
+                result = Version.valueOf(matcher.group()).incrementPatchVersion("SNAPSHOT").toString();
```
Nice use of the `Version` library!

mockitoguy commented on src/main/java/org/shipkit/auto/version/VersionConfig.java:
```diff
@@ -99,6 +99,10 @@ static boolean isSupportedVersion(String version) {
         return version.matches("\\d+\\.\\d+\\.\\d+");
     }
 
+    static boolean isSnapshot(String tag, String tagPrefix) {
```
Let's add javadoc

shestee replied:
@mockitoguy Thanks for your review! As you suggested I refactored isSupportedVersion() and isSnapshot() and covered these methods with unit tests. I also added javadoc.

mockitoguy reviewed (approved):
Nice, please consider my feedback!

shestee committed:
Refactor isSupportedVersion() and isSnapshot()

1. Refactored isSupportedVersion() and isSnapshot() methods with logic moved
from isVersionTag() method which is now removed from TagConvention class.
2. Refactored unit test for isSupportedVersion() and added test for isSnapshot().
3. Added javadoc to isSnapshot() and corrected javadoc of isSupportedVersion()
according to changes provided.

mockitoguy commented on src/test/groovy/org/shipkit/auto/version/VersionConfigTest.groovy:
```diff
@@ -43,17 +44,32 @@ class VersionConfigTest extends TmpFolderSpecification {
-        'x'         | false
-        '1.0'       | false
-        '1.0.0-rc'  | false
+        'x'         | ''         | false
+        'v1.0'      | 'v'        | false
+        'v1.0.0-rc' | 'v'        | false
```
Let's add few more negative cases:
 - good version without tag prefix: `1.2.3`
 - good version with different tag prefix: `ver-1.2.3`

shestee replied:
Good point! I have just added some more cases.

mockitoguy reviewed (approved):
Let's tighten the unit test and ship it!

shestee committed:
Added cases to unit tests
